plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/DefaultJdbcMetadata.java

```diff
@@ -1042,16 +1044,54 @@ public Optional<ConnectorOutputMetadata> finishInsert(ConnectorSession session,
     }
 
     @Override
-    public ColumnHandle getMergeRowIdColumnHandle(ConnectorSession session, ConnectorTableHandle tableHandle)
+    public JdbcColumnHandle getMergeRowIdColumnHandle(ConnectorSession session, ConnectorTableHandle tableHandle)
     {
         verify(!isTableHandleForProcedure(tableHandle), "Not a table reference: %s", tableHandle);
-        // The column is used for row-level merge, which is not supported, but it's required during analysis anyway.
+
+        JdbcTableHandle handle = (JdbcTableHandle) tableHandle;
+        List<RowType.Field> fields = jdbcClient.getPrimaryKeys(session, handle).stream()
+                .map(columnHandle -> new RowType.Field(Optional.of(columnHandle.getColumnName()), columnHandle.getColumnType()))
+                .collect(toImmutableList());
+        if (!fields.isEmpty()) {
+            return new JdbcColumnHandle(
+                    MERGE_ROW_ID,
+                    new JdbcTypeHandle(Types.ROWID, Optional.empty(), Optional.empty(), Optional.empty(), Optional.empty(), Optional.empty()),
+                    RowType.from(fields));
+        }
+
         return new JdbcColumnHandle(
                 MERGE_ROW_ID,
-                new JdbcTypeHandle(Types.BIGINT, Optional.of("bigint"), Optional.empty(), Optional.empty(), Optional.empty(), Optional.empty()),
+                new JdbcTypeHandle(Types.BIGINT, Optional.empty(), Optional.empty(), Optional.empty(), Optional.empty(), Optional.empty()),
                 BIGINT);
     }
 
+    @Override
+    public ConnectorMergeTableHandle beginMerge(ConnectorSession session, ConnectorTableHandle tableHandle, RetryMode retryMode)
+    {
+        JdbcTableHandle handle = (JdbcTableHandle) tableHandle;
+        checkArgument(handle.isNamedRelation(), "Merge target must be named relation table");
+        JdbcTableHandle plainTable = handle.toPlainTableWithoutColumns();
+        JdbcColumnHandle mergeRowIdColumnHandle = getMergeRowIdColumnHandle(session, plainTable);
+        if (jdbcClient.getPrimaryKeys(session, plainTable).isEmpty()) {
+            throw new TrinoException(NOT_SUPPORTED, MODIFYING_ROWS_MESSAGE);
```

Comment:
this can also happen if a specific table doesn't have a primary key - the error should mention this instead of the generic error.

Reply:
We need to add one more method to let metadata know the connector support modifying rows or not.
After that the logic would be: 
- Throws the generic error. the connector not support merge
- Throws the specify error. the connector support merge but the target table doesn't have primary keys. 
Is it that we want?

Reply:
I think let's start with what we have currently. We can improve the error messages in follow-up since it would require changes across lot of classes and make this PR harder to review.

Can you please file an issue in the repo to track this?

Reply:
File https://github.com/trinodb/trino/issues/21132